src/DAT/broadcast.jl

```diff
@@ -17,14 +17,19 @@ function Base.materialize(bc::Broadcast.Broadcasted{XStyle})
     args2 = map(arg -> arg isa Broadcast.Broadcasted ? Base.materialize(arg) : arg, bc.args)
     args2 = map(to_yax, args2)
     # determine output type by calling `eltype` on a dummy function call
-    dummy_args = map(a -> first(a.data), args2)
-    outtype = typeof(bc.f(dummy_args...))
+    intypes = (eltype.(args2)...,)
+    @debug intypes
+    outtypes = Base.return_types(bc.f, intypes)
+    outtype = Union{outtypes...}
+    @debug outtype
     return xmap(XFunction(bc.f; inplace=false), args2..., output=XOutput(; outtype))
 end
 function Base.materialize!(bc::Broadcast.Broadcasted{XStyle})
     args2 = map(arg -> arg isa Broadcast.Broadcasted ? Base.materialize(arg) : arg, bc.args)
     args2 = map(to_yax, args2)
-    dummy_args = map(a -> first(a.data), args2)
-    outtype = typeof(bc.f(dummy_args...))
+        intypes = (eltype.(args2)...,)
+    @debug intypes
+    outtypes = Base.return_types(bc.f, intypes)
+    outtype = Union{outtypes...}
```

Comment:
this doesn't do unions of `Float64`, `Float32` and such, right? I guess promotion applies there. Or things like `Union{Float64, Float64}`, because of the `splat`.

Reply:
Ah yes, you are right we should be doing a promote_type instead of the Union to merge Float32 and Float64 to return Float64

Reply:
```suggestion
    outtype = promote_type{outtypes...}
```